CONTRIBUTING.md

```diff
@@ -1,197 +1,221 @@
 
 ### Seasoned developers
 
-Once you have gotten your feet wet and are more comfortable with the code, you
-can look at the prioritized issues for our next release in our [project boards](https://github.com/rapidsai/cudf/projects).
+Once you have gotten your feet wet and are more comfortable with the code, you can look at the
+prioritized issues for our next release in our
+[project boards](https://github.com/rapidsai/cudf/projects).
 
-> **Pro Tip:** Always look at the release board with the highest number for
-issues to work on. This is where RAPIDS developers also focus their efforts.
+**Note:** Always look at the release board that is
+[currently under development](https://docs.rapids.ai/maintainers) for issues to work on. This is
+where RAPIDS developers also focus their efforts.
 
-Look at the unassigned issues, and find an issue to which you are comfortable
-contributing. Start with _Step 3_ above, commenting on the issue to let
-others know you are working on it. If you have any questions related to the
-implementation of the issue, ask them in the issue instead of the PR.
-
-## Setting Up Your Build Environment
-
-The following instructions are for developers and contributors to cuDF OSS development. These instructions are tested on Linux Ubuntu 16.04 & 18.04. Use these instructions to build cuDF from source and contribute to its development.  Other operating systems may be compatible, but are not currently tested.
+Look at the unassigned issues, and find an issue to which you are comfortable contributing. Start
+with _Step 3_ above, commenting on the issue to let others know you are working on it. If you have
+any questions related to the implementation of the issue, ask them in the issue instead of the PR.
 
+## Setting up your build environment
 
+The following instructions are for developers and contributors to cuDF development. These
+instructions are tested on Ubuntu Linux LTS releases. Use these instructions to build cuDF from
+source and contribute to its development. Other operating systems may be compatible, but are not
+currently tested.
 
 ### General requirements
 
 Compilers:
 
-* `gcc`     version 9.3+
-* `nvcc`    version 11.5+
-* `cmake`   version 3.20.1+
+* `gcc` version 9.3+
+* `nvcc` version 11.5+
+* `cmake` version 3.20.1+
 
 CUDA/GPU:
 
 * CUDA 11.5+
 * NVIDIA driver 450.80.02+
 * Pascal architecture or better
 
-You can obtain CUDA from [https://developer.nvidia.com/cuda-downloads](https://developer.nvidia.com/cuda-downloads).
+You can obtain CUDA from
+[https://developer.nvidia.com/cuda-downloads](https://developer.nvidia.com/cuda-downloads).
+
+### Create the build environment
 
-### Create the build Environment
+- Clone the repository and submodules:
 
-- Clone the repository and submodules
 ```bash
 CUDF_HOME=$(pwd)/cudf
 git clone https://github.com/rapidsai/cudf.git $CUDF_HOME
 cd $CUDF_HOME
 git submodule update --init --remote --recursive
 ```
+
 - Create the conda development environment `cudf_dev`:
+
```

Comment:
This makes it seem like `conda` is required. Many users frequently complain about this and we've gone to great lengths to make it possible to build without conda. 

We should have different sections for "Here's how to build with just cmake" and "here's how to build with a conda environment".

Reply:
Honestly, I have only built libcudf/cudf outside of rapids-compose once. I don't think I'm qualified to write/update these docs but I agree with the idea.

Reply:
I think this is worth putting the effort into as it likely the most valuable change we can make to this document. Building without conda is one of the most frequent user questions/requests we get. 

See 
https://github.com/rapidsai/cudf/issues/2770
https://github.com/rapidsai/cudf/issues/6979

Reply:
@jrhemstad Could we defer this to a later PR? I do think this would be a significant improvement (and may illuminate some challenges with installing from source as mentioned in those issues), but I think it's out of scope for this PR.

Reply:
Minimum I'd like to see for this PR:
- Try building w/o conda and see what happens. If it works, great! Document that this is possible
- If it doesn't:
   -   Add some words to this document that states how a conda environment is currently required to satisfy build dependencies (these could be separated based on building libcudf vs cython/cudf python)
   - List what those dependencies are
   - File an issue to capture what work is needed to avoid requiring a conda environment for building libcudf

Reply:
> * Try building w/o conda and see what happens. If it works, great! Document that this is possible

Building w/o conda is totally possible. We previously had the steps inside a docker file to make this happen for the entire `cudf`, but was removed : https://github.com/rapidsai/cudf/pull/10069/files#diff-64993c0abc780bf9aca0d674016ebd25262ff0c83cb55824cc540455d60b2763

`docker_build/Dockerfile` is the file. Is this close to what you are looking for Jake? Some of the commands _could_ be outdated or unnecessary now but those should be helpful if you are looking to move those steps to here.

Reply:
I deactivated all conda environments (`which python -> /usr/bin/python`) and attempted to run `build.sh`. It failed immediately:

```
-- NO RAPIDS_VERSION for CUDAToolkit
CMake Error at /usr/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find ZLIB (missing: ZLIB_LIBRARY ZLIB_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-3.23/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /usr/share/cmake-3.23/Modules/FindZLIB.cmake:120 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
  build/_deps/rapids-cmake-src/rapids-cmake/find/package.cmake:89 (find_package)
  CMakeLists.txt:125 (rapids_find_package)
```

I estimate it will take me at least 3 hours to figure out how to acquire all the dependencies and build this without conda, and I will probably need to consult with others to make progress here. If you think it's a blocker for this PR, I can work on that further -- I just want to make it clear that it will be a significant time investment and that I feel the PR is otherwise ready. I aimed to scope this PR for eliminating outdated content, standardizing formatting, and making minor updates where appropriate, and this feels like it will require a substantial addition of new content.

Reply:
Could we write the following?

> The build dependencies of libcudf and the cudf Python library are easiest to satisfy through the use of a conda environment. The list of dependencies can be found in `conda/environments/cudf_dev_cuda11.5.yml`. Building from source outside of a conda environment is not currently tested.

Reply:
@galipremsagar Ah! I didn't see your comment while writing my responses above. That's super helpful. The scope/difficulty of documenting this is way lower if I can start from that Dockerfile and update it. In that case, I am much happier about trying to do this in the current PR. 💯

Reply:
> @galipremsagar Ah! I didn't see your comment while writing my responses above. That's super helpful. The scope/difficulty of documenting this is way lower if I can start from that Dockerfile and update it. In that case, I am much happier about trying to do this in the current PR. 💯

Yeah, most of the heavy lifting & testing was done up to a certain previous version of `cudf`. But happy to help if you run into issues using that script now.

Reply:
https://github.com/rapidsai/cudf/pull/10644#discussion_r851330772

This is fine too.

Reply:
Having the following packages would allow you to build cudf C++ library from scratch presuming you have an internet connection, as the rest of the dependencies will be downloaded as needed.

- cuda toolkit 
- cmake ( from https://apt.kitware.com/ )
- build-essential
- g++-9
- libpython3-dev
- libssl-dev 
- libz-dev

Reply:
I think it would be a good idea to maintain a non-conda build instructions, because right after we removed the docker file merlin team reached out with this request: https://github.com/rapidsai/cudf/issues/10301#issuecomment-1041868347

Reply:
Here is a minimal build for c++ ( libcudf )

https://gist.github.com/robertmaynard/93839d326f3b179cd04cbe1b371e229f

Reply:
@jrhemstad I was able to build the minimal Dockerfile that @robertmaynard linked. I committed the list of minimal dependencies in c34658259194e26e6d46563536b55781c5a007eb and this PR is ready for your review. I was able to build without `software-properties-common` so I did not include that one.

Do we need to include a special note about `--cmake-args=\"\-DCUDF_ENABLE_ARROW_S3=OFF\"`? (Or disable that option by default, like the other disabled-by-default I/O features?)

Reply:
Also, we list `gcc 9.3+` in the "General requirements." Do we need to list `build-essential` in the core dependencies as well, or is it reasonable to think that listing a compiler requirement will necessitate the other components of `build-essential`?

Reply:
Our base docker file brings in `build-essential`, but we can list it here so that users know it is required when building from a clean machine ( non-docker ).

Reply:
> Do we need to include a special note about --cmake-args=\"\-DCUDF_ENABLE_ARROW_S3=OFF\"

Good question. I am noticing some difference in cmake search paths in the dockerfile compared to base Ubuntu. We might be able to enable s3 once I figure out what is going wrong.

Reply:
@jlowe would you mind reviewing the section about building w/o conda and verifying that this documentation would be sufficient to reproduce what you've been doing for jar builds?

Reply:
Oops, I put this information in an awkward place. At first I couldn't decide how to organize this information and then I forgot to change it once I made a decision. I am going to put these dependencies between "Create the build environment" and "Build cuDF from source." The instructions will be the same, calling `build.sh`, so we just need to list the dependencies above that.

Reply:
> Do we need to include a special note about `--cmake-args=\"\-DCUDF_ENABLE_ARROW_S3=OFF\"`? (Or disable that option by default, like the other disabled-by-default I/O features?)

So we have a couple of options:

- We can enable arrow by also installing libcurl4-openssl-dev and updating the  `get_arrow.cmake` script to pass `-DAWSSDK_SOURCE=AUTO`
- Or we can document why we are disabling S3 ( less requirements? )

I am hestiant to change the default value of `CUDF_ENABLE_ARROW_S3` as that will be a breaking change to anyone building cudf presuming enabled by default ( users of `build.sh` or people directly calling `cmake` ).

As for why we need to pass `AWSSDK_SOURCE=AUTO` that is required for Arrow's build to work correctly when you don't have arrow already installled. This tells it you are okay with it building the dependency from source when it can't find it locally.

Reply:
Fixed in f674f7f8a1. @jlowe Does this make more sense now?

Reply:
Yep, this makes more sense. Thanks!